### ISSUE TITLE
[suggestion] TransformControls improvements

### ISSUE THREAD

**Issue by gsimone** [react-spring/drei#36]
1. add a `mode` prop that handles the side effect ( control.setMode(mode) )
2. add a `disableControls` (needs a better name) prop that disables controls when interacting with the TransformControls

```jsx
const orbitControls = useRef()
...
<TransformControls
    mode={"scale"}
    disableControls={orbitControls}
>
...
</TransformControls>
<OrbitControls ref={orbitControls} />
```

These are all typical cases and I don't see any reason not to add them

Based on the demo: https://codesandbox.io/s/r3f-drei-transformcontrols-uvqn6?file=/src/index.js:496-526

**Comment by drcmda** [comment]
it has a mode props, its a string. disablecontrols seems like a useful addition! 👍

**Comment by Fasani** [comment]
https://threejs.org/docs/#examples/en/controls/TransformControls has:
```
.enabled : Boolean
Whether or not the controls are enabled.
```

So this should work like this out of the box?

**Comment by gsimone** [comment]
The mode prop only works on first render - and I guess the other props too. To make it props-controlled, this could work - unless we want to check changed props, not sure if there is any real performance issue:

```jsx
useEffect(() => {

   foreach(prop in availableProps) {
       controlsRef.current[prop] = prop
   }

})
```

**Comment by gsimone** [comment]
https://focused-meninsky-e3f9e4.netlify.app/?path=/story/controls-transformcontrols--transform-controls-lock-st 

Here, check the knobs addon, changing the props doesn't work because they are only passed at construction

**Comment by Fasani** [comment]
And if you add the decorator to the export does that change anything?

```
export default {
  title: "Storybook Knobs",
  decorators: [withKnobs]
};
```

**Comment by gsimone** [comment]
The story with the knobs is the second one - line 83 for the decorator -

**Comment by Fasani** [comment]
The version I have locally right now it seems to work. `/Users/fasani/Projects/drei/.storybook/stories/OrbitControls.stories.js`
COmmit hash: `913ebfa4597e06a4590fa2e4e94ba50babadaa4f`
```
export default {
  title: 'Controls/OrbitControls',
  component: OrbitControls,
  decorators: [withKnobs, (storyFn) => <Setup controls={false}>{storyFn()}</Setup>],
}
```
In the version online it looks like you did `OrbitControls.story = {` and put the decorator there but it should be 1 level higher on `OrbitControls`?

**Comment by gsimone** [comment]
I'm sorry, I now realize we are talking about two different things; TransformControls is the one that doesn't change when it receives new props. The story on OrbitControls is fine, I put knobs to change the enableZoom/Pan/Rotate

**Comment by Fasani** [comment]
I feel like it has to be related to the storybook setup, I modified this example online and the arrows do hide/show when changing state plus mode works and in theory it the same conceptually. Just different points of connection (storybook vs react-three-gui)

https://codesandbox.io/s/r3f-drei-transformcontrols-l028w?file=/src/index.js

**Comment by gsimone** [comment]
I'm sorry, the storybook looks like the original one, maybe it didn't save

**Comment by Fasani** [comment]
I pressed save, but still mode is there I guess?

**Comment by gsimone** [comment]
You are right, it does work https://codesandbox.io/s/r3f-drei-transformcontrols-rhvgz?file=/src/index.js I put mode as a prop and it works as expected. Weird, I'll try looking into it.

**Comment by gsimone** [comment]
@drcmda you should probably update the TransformControls sandbox example to use `mode` prop instead of `controls.setMode(mode)`

**Comment by gsimone** [comment]
It was indeed a storybook problem:

```jsx

// the component/boilerplate I want to make a story about
const MyComponent = ({ myProp }) => <div>{myProp && "Hello"}</div> 

// the actual story
export const Story = () => <MyComponent x={boolean('myProp', false) /> 
```

Knobs on Story work while knobs on MyComponent wouldn't. It works in OrbitControl's story because IT IS the story component.
